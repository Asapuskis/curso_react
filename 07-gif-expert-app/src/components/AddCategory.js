import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ( {setCategories} ) => {

    // Creamos un useState para almacenar el valor de la caja de texto y la iniciamos a Kps
    const [inputValue, setInputValue] = useState('Kps');

    // Manejar los cambios de la caja de texto mediante useState.
    const handleInputChange = (e) => {
        setInputValue(e.target);
        console.log('handleInputChange llamado');
    }

    // Manejar el submit del formulario.
    const handleSubmit = (e) => {
        // Mediante preventDefault, evitamos que se refresque toda la página y perder los estados.
        e.preventDefault();

        console.log('handleSubmit', inputValue);

        if (inputValue.trim().length > 2) {
            setCategories( cats => [ inputValue, ...cats,]);
            setInputValue('');
            console.log('Cambios hechos');
        }
    }

    // Devolvemos un formulario para añadir la categoría a la lista.
    return (
        <form onSubmit={handleSubmit}>
            <h2>AddCategory</h2>
            <p>{inputValue}</p>
            <input
                type='text'
                value={inputValue}
                onChange={ handleInputChange }
            />
        </form>
    )
}

// Añadimos el propType requerido para la función setCategories
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
