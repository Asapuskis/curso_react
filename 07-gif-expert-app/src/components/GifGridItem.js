import React from 'react'
import PropTypes from 'prop-types';

// Acordarse de desestructurar image ya que por defecto viene como si fuese una prop, entonces o ponemos prop.image o {image}
export const GifGridItem = ( {image} ) => {
  console.log(`title: ${image.title}`)
  console.log(`url: ${image.url}`)
  return (
    <div className='card animate__animated animate__fadeIn '>
        <p>{image.title}</p>
        <img src={image.url} alt={image.title} />
    </div>
  )
}

GifGridItem.propTypes = {
  image: PropTypes.object.isRequired
}
