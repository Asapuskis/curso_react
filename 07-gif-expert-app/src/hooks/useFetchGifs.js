// Hook custom para hacer la peticiÃ³n de gifs y devolverlos.

import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = ( category ) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {
            getGifs(category)
                .then( imgs => {
                    // Por defecto, esperaremos 500 milis para poder simular la carga de archivos.
                    setTimeout( () => {
                        setState({
                            data: imgs,
                            loading: false
                        });
                    }, 500);
                });
        },
        [category]
    );

    return state;
}
