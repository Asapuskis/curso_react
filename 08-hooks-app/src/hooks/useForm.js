import { useState } from "react"

// Hook para manejar formularios.
export const useForm = ( initialState = {} ) => {

    const [formValues, setFormValues] = useState(initialState);

    const handleInputChange = ({target}) => {
        setFormValues({
            ...formValues,
            [ target.name ]: target.value
        });
    };

    return {
        formValues,
        handleInputChange
    };

}
