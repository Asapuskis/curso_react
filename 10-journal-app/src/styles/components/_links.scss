.link {
    text-decoration: none;
    color: $dark-grey;
    margin-bottom: 10px;
}

.link:hover {
    text-decoration: underline;
}