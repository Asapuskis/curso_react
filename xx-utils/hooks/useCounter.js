import { useState } from 'react'

// Este custom hook sirve para manejar el estado de un contador.
// Recuerda que por defecto, un hook empieza por use
export const useCounter = ( initialState = 10 ) => {

    const [state, setState] = useState(initialState)

    const increment = ( factor = 1 ) => {
        setState( state + factor );
    };

    const decrement = ( factor = 1 ) => {
        setState( state - factor);
    }

    const reset = () => {
        setState( initialState );
    }

    return {
        state,
        increment,
        decrement,
        reset
    };
};
