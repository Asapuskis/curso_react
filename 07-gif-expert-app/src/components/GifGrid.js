import React from 'react'
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ( {category}) => {


    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    // Código para mostrar los gifs utilizando los hooks useState y useEffect.

    // const [images, setImages] = useState([]);

    // // Hook useEffect. Este hook sirve para ejecutar cierto código de manera condicional.
    // // De esta manera evitaremos llamar al código múltiples veces cuando no es necesario recargar los componentes o el navegador.
    // // useEffect tiene dos argumentos:
    // // 1. El código que se va a ejecutar, en este caso getGifs()
    // // 2. Una lista de dependencias que se tienen que cumplir para que se ejecute este código. Si ponemos vacío [] se ejecutará una única vez.
    // // Si colocamos category, se llamará a este hook cuando la variable category se modifique.
    // useEffect( () => {
    //         getGifs(category)
    //             .then( imgs => setImages( imgs ));
    //     },
    //     [ category ]
    // );

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    // Código para mostrar los gifs con un custom Hook creado expresamente para esto:
    // Lo ideal es utilizar custom hooks que llaman a hooks ya existentes y colocarlos aparte en el código, se recomienda crear carpeta hooks

    const {data, loading} = useFetchGifs(category);
    console.log(data);
    console.log(loading);

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    return (
        <>
            <h3 className='animate__animated animate__fadeIn'>{category}</h3>
            {/* Mediante un operador terniario, mostramos loading mientras carga los gifs y nada cuando los haya cargado */}
            { loading ? <p className='animate__animated animate__flash'>loading</p> : null}
            <div className='card-grid animate__animated animate__fadeIn'>
                {
                    data.map( img => (
                        <GifGridItem
                            image={img}
                            key={img.id}
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}