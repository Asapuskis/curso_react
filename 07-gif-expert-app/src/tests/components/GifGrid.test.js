import { shallow } from 'enzyme';
import { GifGrid } from "../../components/GifGrid";
// Mockeamos la función useFetchGifs de esta manera:
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs')

describe('Pruebas en <GifGrid />', () => {
    test('Debe mostrar <GifGrid /> correctamente', () => {
        // Definimos qué va a devolver la función mockeada con mockReturnValue. En este caso, devolvemos el objeto inicializado a vacío.
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        })
        // Utilizaremos shallow para garantizar recoger lo que se renderiza.
        const wrapper = shallow( <GifGrid category='shingeki'/> );
        // toMatchSnapshot genera un archivo .snap. Estos archivos son fotografías de como queda el componente.
        // Previamente hay que instalar enzyme-to-json: npm install --save-dev enzyme-to-json
        expect( wrapper ).toMatchSnapshot();
    });

    test('debe mostrar items cuando se cargan imágenes con useFetchGifs', () => {
        // Mockeamos la función useFetchGifs para simular que devuelve una lista de gifs
        const gifs = [{
            id: 'mockId',
            url: 'https://mocked.jpg',
            title: 'mockTitle'
        }]
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: true
        })
        const wrapper = shallow( <GifGrid category='shingeki'/> );

        expect(wrapper.find('p').exists()).toBe(true);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    })
})