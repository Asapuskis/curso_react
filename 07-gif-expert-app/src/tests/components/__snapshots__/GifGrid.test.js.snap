// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> Debe mostrar <GifGrid /> correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    shingeki
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    loading
  </p>
  <div
    className="card-grid animate__animated animate__fadeIn"
  />
</Fragment>
`;
